{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "10aa16c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Specie_ID</th>\n",
       "      <th>Taxonomy_ID</th>\n",
       "      <th>Assembly_Accession</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>Genome_Source</th>\n",
       "      <th>CODE_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acer negundo</td>\n",
       "      <td>PBXX</td>\n",
       "      <td>4023</td>\n",
       "      <td>GCA_025594385.1</td>\n",
       "      <td>ASM2559438v1</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>PBXX</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Name Specie_ID  Taxonomy_ID Assembly_Accession      Assembly  \\\n",
       "0  Acer negundo      PBXX         4023    GCA_025594385.1  ASM2559438v1   \n",
       "\n",
       "  Genome_Source CODE_ID  \n",
       "0          NCBI    PBXX  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "from Bio import Entrez,SeqIO\n",
    "Entrez.email = \"huangyan8@genomics.cn\"  # 设置你的邮箱地址\n",
    "from ete3 import NCBITaxa, PhyloTree\n",
    "ncbi = NCBITaxa()\n",
    "from collections import Counter\n",
    "# 提供一个物种Taxonomy ID的列表\n",
    "table_path='C:\\\\Users\\\\huangyan8\\\\Desktop\\\\work\\\\2024-09-03 TE_Evolution_Draft\\\\Tables\\\\'\n",
    "fig_path='C:\\\\Users\\\\huangyan8\\\\Desktop\\\\work\\\\2024-09-03 TE_Evolution_Draft\\\\Figures\\\\'\n",
    "samples=pd.read_excel(\"D:\\\\18.TE_Evolution\\\\00.Sample_Information\\\\Plant_558_Genome_Infor_this_project_2024-08-16.xlsx\").fillna(\"\")\n",
    "#samples.to_excel(\"./Tables/Table S1/Genome_Information_Plant294.xlsx\",index=False)\n",
    "samples.to_excel(table_path+'Table S1\\\\Table S1.TE_Evo_558_Species.xlsx',index=False)\n",
    "species_list =samples['Taxonomy_ID'].unique() # 人类、小鼠、大鼠的示例\n",
    "samples['CODE_ID']=samples['Specie_ID']\n",
    "taxid_dic={}\n",
    "for i in range(samples.shape[0]):\n",
    "    taxid_dic[samples.loc[i,'CODE_ID']]=samples.loc[i,'Taxonomy_ID']\n",
    "samples.head(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8aff6c88",
   "metadata": {},
   "source": [
    "### Prepare Plant Groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0951c859",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Group\n",
       "Angiosperms    517\n",
       "Algae           25\n",
       "Mosses           8\n",
       "Ferns            5\n",
       "Gymnosperms      3\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_specie_type1(taxid):\n",
    "    lineage = ncbi.get_lineage(taxid)\n",
    "    #藻类 - Algae\n",
    "    #苔藓类 - Mosses\n",
    "    #蕨类 - Ferns\n",
    "    #裸子植物 - Gymnosperms\n",
    "    #被子植物 - Angiosperms\n",
    "    #for tax in [4410,4418,4405,13714]:\n",
    "    #    if tax in lineage:\n",
    "    #        return 'Other Eudicotyledons'\n",
    "    if 4447 in lineage:\n",
    "        return 'Angiosperms' #Monocots \n",
    "    elif 3398 in lineage:\n",
    "        return 'Angiosperms' #Dicots\n",
    "    elif 71240 in lineage:\n",
    "        return 'Dicots' #'Angiosperms'\n",
    "    elif 1437183 in lineage:\n",
    "        return 'Dicots' #'Angiosperms'\n",
    "    for tax in [131209,2763,1035538,3041,304574] :\n",
    "        if tax in lineage:\n",
    "            return 'Algae'# 藻类\n",
    "    for tax in [3195,3208]:\n",
    "        if tax in lineage:\n",
    "            return 'Mosses' #苔藓\n",
    "    for tax in [3244,3249,241806]:\n",
    "        if tax in lineage:\n",
    "            return 'Ferns' #蕨类\n",
    "    if 58019 in lineage:\n",
    "        return 'Gymnosperms' #裸子植物\n",
    "samples['Group']=samples['Taxonomy_ID'].apply(get_specie_type1)\n",
    "samples['Group'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8aee0aa",
   "metadata": {},
   "source": [
    "### Prepare Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d6b6b8c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_path='D:\\\\18.TE_Evolution\\\\12.Analysis\\\\2024-08-16 TE Evolution\\\\Part3.Feature_Selection\\\\'\n",
    "anno_path='D:\\\\18.TE_Evolution\\\\03.Protein_Annotation\\\\'\n",
    "kegg_path=anno_path+\"KEGG.ko\\\\\"\n",
    "go_path=anno_path+\"GO\\\\\"\n",
    "PlantTF_path=anno_path+'PlantTF_Family\\\\'\n",
    "FLOR_path=anno_path+'FLOR_Family\\\\'\n",
    "terms_dic={}\n",
    "gene_set={}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "61c42d41",
   "metadata": {},
   "outputs": [],
   "source": [
    "# FLOR\n",
    "for i in range(samples.shape[0]):\n",
    "    specie_id=samples.loc[i,'CODE_ID']\n",
    "    df=pd.read_csv(FLOR_path+specie_id+\"_FLOR.csv\",sep='\\t').dropna()\n",
    "    df['Terms']=df['FLOR_Family'].apply(lambda x:x.split(\":\")[0])\n",
    "    if specie_id not in terms_dic:\n",
    "        terms_dic[specie_id]={}\n",
    "        gene_set[specie_id]={}\n",
    "    for j in range(df.shape[0]):\n",
    "        term=df.loc[j,'Terms']\n",
    "        if term in terms_dic[specie_id]:\n",
    "            terms_dic[specie_id][term]+=1\n",
    "            #gene_set[specie_id][term].append(df.loc[j,'Gene'])\n",
    "        else:\n",
    "            terms_dic[specie_id][term]=1\n",
    "            #gene_set[specie_id][term]=[df.loc[j,'Gene']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "9ee1de28",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PlantTF\n",
    "for i in range(samples.shape[0]):\n",
    "    specie_id=samples.loc[i,'CODE_ID']\n",
    "    df=pd.read_csv(PlantTF_path+specie_id+\"_PlantTF.csv\",sep='\\t').dropna()\n",
    "    df['Terms']=df['TF_Family'].apply(lambda x:x.split(\":\")[0])\n",
    "    if specie_id not in terms_dic:\n",
    "        terms_dic[specie_id]={}\n",
    "        gene_set[specie_id]={}\n",
    "    for j in range(df.shape[0]):\n",
    "        term=df.loc[j,'Terms']\n",
    "        if term in terms_dic[specie_id]:\n",
    "            terms_dic[specie_id][term]+=1\n",
    "            #gene_set[specie_id][term].append(df.loc[j,'Gene'])\n",
    "        else:\n",
    "            terms_dic[specie_id][term]=1\n",
    "            #gene_set[specie_id][term]=[df.loc[j,'Gene']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "07619114",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(samples.shape[0]):\n",
    "    specie_id=samples.loc[i,'CODE_ID']\n",
    "    df=pd.read_csv(kegg_path+specie_id+\".ko.csv\").dropna().reset_index(drop=True)\n",
    "    #df=pd.read_csv(go_path+specie_id+\".wego.csv\").dropna()\n",
    "    if specie_id not in terms_dic:\n",
    "        terms_dic[specie_id]={}\n",
    "    for j in range(df.shape[0]):\n",
    "        terms=df.loc[j,'KEGG_Terms']\n",
    "        if '\\t' in terms:\n",
    "            terms=terms.split(\"\\t\")\n",
    "        elif ' 'in terms:\n",
    "            terms=terms.split(\" \")\n",
    "        else:\n",
    "            terms=[terms]\n",
    "        for term in terms:\n",
    "            if term in terms_dic[specie_id]:\n",
    "                terms_dic[specie_id][term]+=1\n",
    "                #gene_set[specie_id][term].append(df.loc[j,'Gene'])\n",
    "            else:\n",
    "                terms_dic[specie_id][term]=1\n",
    "                #gene_set[specie_id][term]=[df.loc[j,'Gene']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "1508d3a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "50\n",
      "100\n",
      "150\n",
      "200\n",
      "250\n",
      "300\n",
      "350\n",
      "400\n",
      "450\n",
      "500\n",
      "550\n"
     ]
    }
   ],
   "source": [
    "for i in range(samples.shape[0]):\n",
    "    if i%50==0:\n",
    "        print(i)\n",
    "    specie_id=samples.loc[i,'CODE_ID']\n",
    "    df=pd.read_csv(go_path+specie_id+\".wego.csv\").dropna().reset_index(drop=True)\n",
    "    if specie_id not in terms_dic:\n",
    "        terms_dic[specie_id]={}\n",
    "    for j in range(df.shape[0]):\n",
    "        terms=df.loc[j,'Go_Terms']\n",
    "        if '\\t' in terms:\n",
    "            terms=terms.split(\"\\t\")\n",
    "        elif ' 'in terms:\n",
    "            terms=terms.split(\" \")\n",
    "        else:\n",
    "            terms=[terms]\n",
    "        for term in terms:\n",
    "            if term in terms_dic[specie_id]:\n",
    "                terms_dic[specie_id][term]+=1\n",
    "            else:\n",
    "                terms_dic[specie_id][term]=1\n",
    "            #if term in gene_set[specie_id]:\n",
    "            #    gene_set[specie_id][term].append(df.loc[j,'Gene'])\n",
    "            #else:\n",
    "            #    gene_set[specie_id][term]=[df.loc[j,'Gene']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f083b317",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "50\n",
      "100\n",
      "150\n",
      "200\n",
      "250\n",
      "300\n",
      "350\n",
      "400\n",
      "450\n",
      "500\n",
      "550\n"
     ]
    }
   ],
   "source": [
    "#ncRNA\n",
    "ncRNA_path='D:\\\\18.TE_Evolution\\\\06.Bed\\\\ncRNA_Bed\\\\'\n",
    "for i in range(samples.shape[0]):\n",
    "    if i%50==0:\n",
    "        print(i)\n",
    "    specie_id=samples.loc[i,'CODE_ID']\n",
    "    if specie_id+\"_ncRNA.bed\" in os.listdir(ncRNA_path):\n",
    "        df=pd.read_csv(ncRNA_path+specie_id+\"_ncRNA.bed\",sep='\\t').dropna().reset_index(drop=True)\n",
    "        if specie_id not in terms_dic:\n",
    "            terms_dic[specie_id]={}\n",
    "        for f,f_df in df.groupby('ncRNA_Family'):\n",
    "            terms_dic[specie_id][f]=f_df.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4ca04959",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(558, 72537)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FLOR_0</th>\n",
       "      <th>FLOR_1</th>\n",
       "      <th>FLOR_2</th>\n",
       "      <th>FLOR_3</th>\n",
       "      <th>FLOR_5</th>\n",
       "      <th>FLOR_7</th>\n",
       "      <th>FLOR_39</th>\n",
       "      <th>FLOR_10</th>\n",
       "      <th>FLOR_11</th>\n",
       "      <th>FLOR_12</th>\n",
       "      <th>...</th>\n",
       "      <th>K48044</th>\n",
       "      <th>K48045</th>\n",
       "      <th>K48046</th>\n",
       "      <th>K48047</th>\n",
       "      <th>K48048</th>\n",
       "      <th>K48049</th>\n",
       "      <th>K48050</th>\n",
       "      <th>K48051</th>\n",
       "      <th>RF02020</th>\n",
       "      <th>RF02869</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PBXX</th>\n",
       "      <td>99</td>\n",
       "      <td>40</td>\n",
       "      <td>1141</td>\n",
       "      <td>157.0</td>\n",
       "      <td>183.0</td>\n",
       "      <td>160.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>122</td>\n",
       "      <td>8.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>QIBM</th>\n",
       "      <td>118</td>\n",
       "      <td>42</td>\n",
       "      <td>1214</td>\n",
       "      <td>147.0</td>\n",
       "      <td>208.0</td>\n",
       "      <td>212.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>135</td>\n",
       "      <td>10.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BUKK</th>\n",
       "      <td>209</td>\n",
       "      <td>52</td>\n",
       "      <td>1769</td>\n",
       "      <td>177.0</td>\n",
       "      <td>308.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>197</td>\n",
       "      <td>14.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GJNO</th>\n",
       "      <td>110</td>\n",
       "      <td>27</td>\n",
       "      <td>988</td>\n",
       "      <td>86.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>92.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>119</td>\n",
       "      <td>11.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WWFJ</th>\n",
       "      <td>199</td>\n",
       "      <td>63</td>\n",
       "      <td>1673</td>\n",
       "      <td>160.0</td>\n",
       "      <td>382.0</td>\n",
       "      <td>158.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>222</td>\n",
       "      <td>12.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 72537 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      FLOR_0  FLOR_1  FLOR_2  FLOR_3  FLOR_5  FLOR_7  FLOR_39  FLOR_10  \\\n",
       "PBXX      99      40    1141   157.0   183.0   160.0     67.0      9.0   \n",
       "QIBM     118      42    1214   147.0   208.0   212.0     78.0      9.0   \n",
       "BUKK     209      52    1769   177.0   308.0   147.0    114.0     15.0   \n",
       "GJNO     110      27     988    86.0   157.0    92.0     58.0      8.0   \n",
       "WWFJ     199      63    1673   160.0   382.0   158.0    133.0     20.0   \n",
       "\n",
       "      FLOR_11  FLOR_12  ...  K48044  K48045  K48046  K48047  K48048  K48049  \\\n",
       "PBXX      122      8.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "QIBM      135     10.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "BUKK      197     14.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "GJNO      119     11.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "WWFJ      222     12.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "      K48050  K48051  RF02020  RF02869  \n",
       "PBXX     0.0     0.0      0.0      0.0  \n",
       "QIBM     0.0     0.0      0.0      0.0  \n",
       "BUKK     0.0     0.0      0.0      0.0  \n",
       "GJNO     0.0     0.0      0.0      0.0  \n",
       "WWFJ     0.0     0.0      0.0      0.0  \n",
       "\n",
       "[5 rows x 72537 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Feature=pd.DataFrame.from_dict(terms_dic,orient='index').fillna(0)#.reset_index()\n",
    "Feature.to_csv(\"D:\\\\18.TE_Evolution\\\\12.Analysis\\\\2024-08-16 TE Evolution\\\\Part3.Feature_Selection\\\\Plant558_Features.csv\")\n",
    "print(Feature.shape)\n",
    "Feature.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "732760dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(558, 72538)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>FLOR_0</th>\n",
       "      <th>FLOR_1</th>\n",
       "      <th>FLOR_2</th>\n",
       "      <th>FLOR_3</th>\n",
       "      <th>FLOR_5</th>\n",
       "      <th>FLOR_7</th>\n",
       "      <th>FLOR_39</th>\n",
       "      <th>FLOR_10</th>\n",
       "      <th>FLOR_11</th>\n",
       "      <th>...</th>\n",
       "      <th>K48044</th>\n",
       "      <th>K48045</th>\n",
       "      <th>K48046</th>\n",
       "      <th>K48047</th>\n",
       "      <th>K48048</th>\n",
       "      <th>K48049</th>\n",
       "      <th>K48050</th>\n",
       "      <th>K48051</th>\n",
       "      <th>RF02020</th>\n",
       "      <th>RF02869</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PBXX</td>\n",
       "      <td>99</td>\n",
       "      <td>40</td>\n",
       "      <td>1141</td>\n",
       "      <td>157.0</td>\n",
       "      <td>183.0</td>\n",
       "      <td>160.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>122</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>QIBM</td>\n",
       "      <td>118</td>\n",
       "      <td>42</td>\n",
       "      <td>1214</td>\n",
       "      <td>147.0</td>\n",
       "      <td>208.0</td>\n",
       "      <td>212.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>135</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BUKK</td>\n",
       "      <td>209</td>\n",
       "      <td>52</td>\n",
       "      <td>1769</td>\n",
       "      <td>177.0</td>\n",
       "      <td>308.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>197</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GJNO</td>\n",
       "      <td>110</td>\n",
       "      <td>27</td>\n",
       "      <td>988</td>\n",
       "      <td>86.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>92.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>119</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WWFJ</td>\n",
       "      <td>199</td>\n",
       "      <td>63</td>\n",
       "      <td>1673</td>\n",
       "      <td>160.0</td>\n",
       "      <td>382.0</td>\n",
       "      <td>158.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>222</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 72538 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Unnamed: 0  FLOR_0  FLOR_1  FLOR_2  FLOR_3  FLOR_5  FLOR_7  FLOR_39  \\\n",
       "0       PBXX      99      40    1141   157.0   183.0   160.0     67.0   \n",
       "1       QIBM     118      42    1214   147.0   208.0   212.0     78.0   \n",
       "2       BUKK     209      52    1769   177.0   308.0   147.0    114.0   \n",
       "3       GJNO     110      27     988    86.0   157.0    92.0     58.0   \n",
       "4       WWFJ     199      63    1673   160.0   382.0   158.0    133.0   \n",
       "\n",
       "   FLOR_10  FLOR_11  ...  K48044  K48045  K48046  K48047  K48048  K48049  \\\n",
       "0      9.0      122  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "1      9.0      135  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "2     15.0      197  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "3      8.0      119  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "4     20.0      222  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "   K48050  K48051  RF02020  RF02869  \n",
       "0     0.0     0.0      0.0      0.0  \n",
       "1     0.0     0.0      0.0      0.0  \n",
       "2     0.0     0.0      0.0      0.0  \n",
       "3     0.0     0.0      0.0      0.0  \n",
       "4     0.0     0.0      0.0      0.0  \n",
       "\n",
       "[5 rows x 72538 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Feature=pd.read_csv(\"D:\\\\18.TE_Evolution\\\\12.Analysis\\\\2024-08-16 TE Evolution\\\\Part3.Feature_Selection\\\\Plant558_Features.csv\")\n",
    "Feature=Feature.set_index(\"Unnamed: 0\")\n",
    "print(Feature.shape)\n",
    "Feature.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "4ce11301",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FLOR (21, 10)\n",
      "GO (1110, 12)\n",
      "KEGG (1565, 10)\n",
      "PlantTF (73, 10)\n",
      "ncRNA (36, 10)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "2805"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_terms=[]\n",
    "title_dic={}\n",
    "for f_type in ['FLOR','GO',\"KEGG\",\"PlantTF\",'ncRNA']:#\n",
    "    df=pd.read_excel(\"D:\\\\18.TE_Evolution\\\\12.Analysis\\\\2024-08-16 TE Evolution\\\\Part1.Family_Enrichment\\\\\"+f_type+\"_Dom_and_TE_DupGenes_Enrichment.xlsx\")\n",
    "    #df['Term']=df['index']\n",
    "    df['LogP']=df['Adjusted P-value'].apply(lambda x:-np.log10(x))\n",
    "    df=df[df['LogP']>=2].reset_index(drop=True)\n",
    "    print(f_type,df.shape)\n",
    "    for i in range(df.shape[0]):\n",
    "        title_dic[df.loc[i,'Term']]=df.loc[i,'Title']\n",
    "    sig_terms+=df['Term'].tolist()\n",
    "len(sig_terms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "561d7e75",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Angiosperms', 'Algae', 'Mosses', 'Ferns', 'Gymnosperms'], dtype='object', name='Group')"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples['Group'].value_counts().index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "387a5f6e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FLOR_0</th>\n",
       "      <th>FLOR_1</th>\n",
       "      <th>FLOR_2</th>\n",
       "      <th>FLOR_3</th>\n",
       "      <th>FLOR_5</th>\n",
       "      <th>FLOR_7</th>\n",
       "      <th>FLOR_39</th>\n",
       "      <th>FLOR_10</th>\n",
       "      <th>FLOR_11</th>\n",
       "      <th>FLOR_12</th>\n",
       "      <th>...</th>\n",
       "      <th>K48044</th>\n",
       "      <th>K48045</th>\n",
       "      <th>K48046</th>\n",
       "      <th>K48047</th>\n",
       "      <th>K48048</th>\n",
       "      <th>K48049</th>\n",
       "      <th>K48050</th>\n",
       "      <th>K48051</th>\n",
       "      <th>RF02020</th>\n",
       "      <th>RF02869</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PBXX</th>\n",
       "      <td>99</td>\n",
       "      <td>40</td>\n",
       "      <td>1141</td>\n",
       "      <td>157.0</td>\n",
       "      <td>183.0</td>\n",
       "      <td>160.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>122</td>\n",
       "      <td>8.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 72537 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      FLOR_0  FLOR_1  FLOR_2  FLOR_3  FLOR_5  FLOR_7  FLOR_39  FLOR_10  \\\n",
       "PBXX      99      40    1141   157.0   183.0   160.0     67.0      9.0   \n",
       "\n",
       "      FLOR_11  FLOR_12  ...  K48044  K48045  K48046  K48047  K48048  K48049  \\\n",
       "PBXX      122      8.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "      K48050  K48051  RF02020  RF02869  \n",
       "PBXX     0.0     0.0      0.0      0.0  \n",
       "\n",
       "[1 rows x 72537 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Feature.head(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df5d3947",
   "metadata": {},
   "source": [
    "### Get Groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e5bace6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_specie_type1(taxid):\n",
    "    #'Algae', 'Mosses', 'Ferns', 'Gymnosperms','Angiosperms'\n",
    "    lineage = ncbi.get_lineage(taxid)\n",
    "    #藻类 - Algae\n",
    "    #苔藓类 - Mosses\n",
    "    #蕨类 - Ferns\n",
    "    #裸子植物 - Gymnosperms\n",
    "    #被子植物 - Angiosperms\n",
    "    #for tax in [4410,4418,4405,13714]:\n",
    "    #    if tax in lineage:\n",
    "    #        return 'Other Eudicotyledons'\n",
    "    if 4447 in lineage:\n",
    "        return 'Angiosperms' #Monocots \n",
    "    elif 3398 in lineage:\n",
    "        return 'Angiosperms' #Dicots\n",
    "    elif 71240 in lineage:\n",
    "        return 'Dicots' #'Angiosperms'\n",
    "    elif 1437183 in lineage:\n",
    "        return 'Dicots' #'Angiosperms'\n",
    "    for tax in [131209,2763,1035538,3041,304574] :\n",
    "        if tax in lineage:\n",
    "            return 'Algae'# 藻类\n",
    "    for tax in [3195,3208]:\n",
    "        if tax in lineage:\n",
    "            return 'Mosses' #苔藓\n",
    "    for tax in [3244,3249,241806]:\n",
    "        if tax in lineage:\n",
    "            return 'Ferns' #蕨类\n",
    "    if 58019 in lineage:\n",
    "        return 'Gymnosperms' #裸子植物\n",
    "samples['Group']=samples['Taxonomy_ID'].apply(get_specie_type1)\n",
    "samples['Group'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "5c36bf82",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples=samples[samples['Group']=='Angiosperms']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "18a5fffe",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-121-8e3b1da7dd96>:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  samples['Group']=samples['Taxonomy_ID'].apply(get_specie_type2)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Group\n",
       "Dicots      387\n",
       "Monocots    130\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_specie_type2(taxid):\n",
    "    # 'Dicots','Monocots'\n",
    "    lineage = ncbi.get_lineage(taxid)\n",
    "    #藻类 - Algae\n",
    "    #苔藓类 - Mosses\n",
    "    #蕨类 - Ferns\n",
    "    #裸子植物 - Gymnosperms\n",
    "    #被子植物 - Angiosperms\n",
    "    #for tax in [4410,4418,4405,13714]:\n",
    "    #    if tax in lineage:\n",
    "    #        return 'Other Eudicotyledons'\n",
    "    if 4447 in lineage:\n",
    "        return 'Monocots'\n",
    "    elif 3398 in lineage:\n",
    "        return 'Dicots' #\n",
    "    else:\n",
    "        return 'Other'\n",
    "samples['Group']=samples['Taxonomy_ID'].apply(get_specie_type2)\n",
    "samples['Group'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "b4c6f29e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(130, 8)"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples=pd.read_excel(\"D:\\\\18.TE_Evolution\\\\00.Sample_Information\\\\Plant_558_Genome_Infor_this_project_2024-08-16.xlsx\").fillna(\"\")\n",
    "#samples.to_excel(\"./Tables/Table S1/Genome_Information_Plant294.xlsx\",index=False)\n",
    "samples.to_excel(table_path+'Table S1\\\\Table S1.TE_Evo_558_Species.xlsx',index=False)\n",
    "species_list =samples['Taxonomy_ID'].unique() # 人类、小鼠、大鼠的示例\n",
    "samples['CODE_ID']=samples['Specie_ID']\n",
    "samples['Group']=samples['Taxonomy_ID'].apply(get_specie_type2)\n",
    "samples=samples[samples['Group']=='Monocots']\n",
    "samples.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "baa36c6c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3962317, 4)\n"
     ]
    }
   ],
   "source": [
    "tax_path='D:\\\\18.TE_Evolution\\\\00.Sample_Information\\\\Genomes_Info\\\\taxdmp\\\\'\n",
    "tax=pd.read_csv(tax_path+\"nodes.dmp\",sep='\\t').fillna(\"\")\n",
    "use_col=[]\n",
    "\n",
    "for col in tax.columns:\n",
    "    #print(col,tax.loc[0,col])\n",
    "    if tax.loc[0,col]!=\"|\":\n",
    "        use_col.append(col)\n",
    "tax=tax[use_col]\n",
    "tax.columns=[\"tax_id\",'parent tax_id','rank','embl code','division id','inherited div flag','genetic code id','inherited GC flag',\n",
    "            'mitochondrial genetic code id','inherited MGC flag','GenBank hidden flag','hidden subtree root flag','comments']\n",
    "names=pd.read_csv(tax_path+\"names.dmp\",sep='\\t').fillna(\"\")\n",
    "use_col=[]\n",
    "for col in names.columns:\n",
    "    if names.loc[0,col]!=\"|\":\n",
    "        use_col.append(col)\n",
    "names=names[use_col]\n",
    "names.columns=['tax_id','name_txt','unique_name','name class']\n",
    "print(names.shape)\n",
    "taxid_dic={}\n",
    "for taxon_id in names[\"tax_id\"].unique():\n",
    "    taxid_dic[taxon_id]={}\n",
    "    for c in ['scientific name', 'blast name', 'genbank common name', 'in-part',\n",
    "       'authority', 'type material', 'equivalent name', 'includes',\n",
    "       'synonym', 'common name', 'acronym', 'genbank acronym']:\n",
    "        taxid_dic[taxon_id][c]=\"\"\n",
    "for i in range(names.shape[0]):\n",
    "    taxon_id=names.loc[i,\"tax_id\"]\n",
    "    taxid_dic[taxon_id][names.loc[i,'name class']]=names.loc[i,\"name_txt\"]\n",
    "Plant=tax[tax['division id']==4].reset_index(drop=True)\n",
    "Rank_dic={}\n",
    "for i in range(Plant.shape[0]):\n",
    "    Rank_dic[Plant.loc[i,\"tax_id\"]]=Plant.loc[i,'rank']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "2d560fa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_tree(tax_id):\n",
    "    find=True\n",
    "    Parents=[]\n",
    "    while find==True:\n",
    "        df0=Plant[Plant['tax_id']==tax_id]\n",
    "        if df0.shape[0]>0:\n",
    "            tax_id=df0['parent tax_id'].tolist()[0]\n",
    "            if tax_id in Rank_dic:\n",
    "                Parents.append(str(tax_id)+\":\"+Rank_dic[tax_id]+\":\"+taxid_dic[tax_id]['scientific name'])\n",
    "        else:\n",
    "            find=False\n",
    "    return \" | \".join(Parents[::-1])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "19ab06d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Family\n",
       "4479:family:Poaceae          100\n",
       "4747:family:Orchidaceae        5\n",
       "4609:family:Cyperaceae         4\n",
       "4710:family:Arecaceae          3\n",
       "4671:family:Dioscoreaceae      3\n",
       "4637:family:Musaceae           3\n",
       "42228:family:Acoraceae         2\n",
       "4613:family:Bromeliaceae       2\n",
       "4454:family:Araceae            2\n",
       "27254:family:Zosteraceae       2\n",
       "4626:family:Cannaceae          1\n",
       "14101:family:Juncaceae         1\n",
       "4642:family:Zingiberaceae      1\n",
       "40552:family:Asparagaceae      1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_specie_type3(taxid):\n",
    "    # 'Dicots','Monocots'\n",
    "    lineage = get_tree(taxid)#ncbi.get_lineage(taxid)\n",
    "    def get_family(x):\n",
    "        if x ==\"\":\n",
    "            return \"\"\n",
    "        values=x.split(\" | \")\n",
    "        Keywords=[]\n",
    "        for value in values:\n",
    "            kw=value.split(\":\")[1]\n",
    "            if kw in [\"family\"]:\n",
    "                Keywords.append(value)\n",
    "        return \" | \".join(Keywords)\n",
    "    family=get_family(lineage)\n",
    "    return family\n",
    "samples['Family']=samples['Taxonomy_ID'].apply(get_specie_type3)\n",
    "samples['Family'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "id": "413b76cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Group\n",
       "Poaceae          100\n",
       "Orchidaceae        5\n",
       "Cyperaceae         4\n",
       "Arecaceae          3\n",
       "Dioscoreaceae      3\n",
       "Musaceae           3\n",
       "Acoraceae          2\n",
       "Bromeliaceae       2\n",
       "Araceae            2\n",
       "Zosteraceae        2\n",
       "Cannaceae          1\n",
       "Juncaceae          1\n",
       "Zingiberaceae      1\n",
       "Asparagaceae       1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples['Group']=samples['Family'].apply(lambda x:x.split(\":\")[-1].split(\" \")[0])\n",
    "samples['Group'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "8de8ab32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Fabaceae', 'Brassicaceae', 'Solanaceae', 'Rosaceae', 'Asteraceae',\n",
       "       'Salicaceae', 'Cucurbitaceae', 'Malvaceae', 'Euphorbiaceae',\n",
       "       'Sapindaceae'],\n",
       "      dtype='object', name='Group')"
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2=pd.DataFrame(samples['Group'].value_counts())\n",
    "df2.head(10).index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "b66d0d60",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples=samples[samples['Group']=='Poaceae']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "b187a5d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Group\n",
       "Oryzoideae       42\n",
       "Panicoideae      40\n",
       "Pooideae          9\n",
       "Chloridoideae     7\n",
       "Bambusoideae      1\n",
       "Arundinoideae     1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_specie_type4(taxid):\n",
    "    # 'Dicots','Monocots'\n",
    "    lineage = get_tree(taxid)#ncbi.get_lineage(taxid)\n",
    "    def get_subfamily(x):\n",
    "        if x ==\"\":\n",
    "            return \"\"\n",
    "        values=x.split(\" | \")\n",
    "        Keywords=[]\n",
    "        for value in values:\n",
    "            kw=value.split(\":\")[1]\n",
    "            if kw in [\"subfamily\"]:\n",
    "                Keywords.append(value)\n",
    "        return \" | \".join(Keywords)\n",
    "    family=get_subfamily(lineage)\n",
    "    return family\n",
    "samples['Sub_Family']=samples['Taxonomy_ID'].apply(get_specie_type4)\n",
    "samples['Sub_Family'].value_counts()\n",
    "samples['Group']=samples['Sub_Family'].apply(lambda x:x.split(\":\")[2].split(\" \")[0])\n",
    "samples['Group'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "9887320f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Group\n",
       "Other    517\n",
       "Oryza     41\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples=pd.read_excel(\"D:\\\\18.TE_Evolution\\\\00.Sample_Information\\\\Plant_558_Genome_Infor_this_project_2024-08-16.xlsx\").fillna(\"\")\n",
    "#samples.to_excel(\"./Tables/Table S1/Genome_Information_Plant294.xlsx\",index=False)\n",
    "samples.to_excel(table_path+'Table S1\\\\Table S1.TE_Evo_558_Species.xlsx',index=False)\n",
    "species_list =samples['Taxonomy_ID'].unique() # 人类、小鼠、大鼠的示例\n",
    "samples['CODE_ID']=samples['Specie_ID']\n",
    "samples['Group']=samples['Name'].apply(lambda x: 'Oryza' if 'Oryza' in x else 'Other')\n",
    "samples['Group'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "fb731882",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Specie_ID</th>\n",
       "      <th>Taxonomy_ID</th>\n",
       "      <th>Assembly_Accession</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>Genome_Source</th>\n",
       "      <th>CODE_ID</th>\n",
       "      <th>Group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>347</th>\n",
       "      <td>Oryza sativa 02428</td>\n",
       "      <td>ILWU</td>\n",
       "      <td>4530</td>\n",
       "      <td>Oryza sativa 02428</td>\n",
       "      <td>02428</td>\n",
       "      <td>Rice Resource Center</td>\n",
       "      <td>ILWU</td>\n",
       "      <td>Oryza</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>348</th>\n",
       "      <td>Oryza sativa 9311</td>\n",
       "      <td>OGXT</td>\n",
       "      <td>4530</td>\n",
       "      <td>Oryza sativa 9311</td>\n",
       "      <td>9311</td>\n",
       "      <td>Rice Resource Center</td>\n",
       "      <td>OGXT</td>\n",
       "      <td>Oryza</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>Oryza sativa Basmati1</td>\n",
       "      <td>MPZB</td>\n",
       "      <td>4530</td>\n",
       "      <td>Oryza sativa Basmati1</td>\n",
       "      <td>Basmati1</td>\n",
       "      <td>Rice Resource Center</td>\n",
       "      <td>MPZB</td>\n",
       "      <td>Oryza</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>Oryza sativa CN1</td>\n",
       "      <td>LMQO</td>\n",
       "      <td>4530</td>\n",
       "      <td>Oryza sativa CN1</td>\n",
       "      <td>CN1</td>\n",
       "      <td>Rice Resource Center</td>\n",
       "      <td>LMQO</td>\n",
       "      <td>Oryza</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>Oryza sativa D62</td>\n",
       "      <td>BZJI</td>\n",
       "      <td>4530</td>\n",
       "      <td>Oryza sativa D62</td>\n",
       "      <td>D62</td>\n",
       "      <td>Rice Resource Center</td>\n",
       "      <td>BZJI</td>\n",
       "      <td>Oryza</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Name Specie_ID  Taxonomy_ID     Assembly_Accession  \\\n",
       "347     Oryza sativa 02428      ILWU         4530     Oryza sativa 02428   \n",
       "348      Oryza sativa 9311      OGXT         4530      Oryza sativa 9311   \n",
       "349  Oryza sativa Basmati1      MPZB         4530  Oryza sativa Basmati1   \n",
       "351       Oryza sativa CN1      LMQO         4530       Oryza sativa CN1   \n",
       "352       Oryza sativa D62      BZJI         4530       Oryza sativa D62   \n",
       "\n",
       "     Assembly         Genome_Source CODE_ID  Group  \n",
       "347     02428  Rice Resource Center    ILWU  Oryza  \n",
       "348      9311  Rice Resource Center    OGXT  Oryza  \n",
       "349  Basmati1  Rice Resource Center    MPZB  Oryza  \n",
       "351       CN1  Rice Resource Center    LMQO  Oryza  \n",
       "352       D62  Rice Resource Center    BZJI  Oryza  "
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples=samples[samples['Group']=='Oryza']\n",
    "samples[samples['Taxonomy_ID']==4530].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "87ff56b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Group\n",
       "Oryza sativa    30\n",
       "Other Oryza     11\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples['Group']=samples['Taxonomy_ID'].apply(lambda x: 'Oryza sativa' if x==4530 else 'Other Oryza')\n",
    "samples['Group'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f1e3776",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6920ca96",
   "metadata": {},
   "source": [
    "### Group Feature Selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "34d214a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.feature_selection import SelectFromModel\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import pandas as pd\n",
    "import scipy.stats as stats\n",
    "scaler = StandardScaler()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "031e6305",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Brassicaceae (203, 12)\n",
      "Fabaceae (32, 12)\n",
      "Solanaceae (49, 12)\n",
      "Rosaceae (17, 12)\n",
      "Asteraceae (16, 12)\n",
      "Salicaceae (25, 12)\n",
      "Cucurbitaceae (2, 12)\n"
     ]
    }
   ],
   "source": [
    "families=['Brassicaceae','Fabaceae', 'Solanaceae', 'Rosaceae', 'Asteraceae',\n",
    "       'Salicaceae', 'Cucurbitaceae']\n",
    "for group in families:\n",
    "    other_groups=[]\n",
    "    for group2 in families:\n",
    "        if group2!=group:\n",
    "            other_groups.append(group2)\n",
    "    R=[]\n",
    "    for group2 in other_groups:\n",
    "        two_groups=[group,group2]\n",
    "        group_idxs=samples[samples['Group']==group]['CODE_ID'].tolist()\n",
    "        other_idxs=samples[samples['Group']==group2]['CODE_ID'].tolist()\n",
    "        data=Feature.loc[group_idxs+other_idxs,:]\n",
    "        X=data[sig_terms].values \n",
    "        #X=scaler.fit_transform(X)\n",
    "        y=[]\n",
    "        for idx in group_idxs:\n",
    "            y.append(1)\n",
    "        for idx in other_idxs:\n",
    "            y.append(0)\n",
    "        y=pd.Series(y)\n",
    "        clf = RandomForestClassifier(n_estimators=1000)\n",
    "        clf.fit(X, y)\n",
    "        importances = clf.feature_importances_\n",
    "        feature_importance_df = pd.DataFrame({'Feature': sig_terms, \"Importance\": importances})\n",
    "        feature_importance_df['Group_Pair']=group+\":\"+group2\n",
    "        R.append(feature_importance_df)\n",
    "    R=pd.concat(R)\n",
    "    R=R[R['Importance']>0]\n",
    "    c=Counter(R['Feature'].tolist())\n",
    "    R['Sig_Count']=R['Feature'].apply(lambda x:c[x])\n",
    "    \n",
    "    df2=pd.DataFrame(R.groupby(\"Feature\")['Importance'].sum()).reset_index()\n",
    "    df2=df2.sort_values('Importance',ascending=False)\n",
    "    df2['Title']=df2['Feature'].apply(lambda x:title_dic[x])\n",
    "    df2=df2.reset_index(drop=True)\n",
    "    df2[group+\"_Score\"]=0\n",
    "    df2[group+\"_SigScore\"]=0\n",
    "    for i in range(df2.shape[0]):\n",
    "        feature=df2.loc[i,'Feature']\n",
    "        group_values=Feature.loc[group_idxs,feature].values\n",
    "        df2.loc[i,group]=round(Feature.loc[group_idxs,feature].values.mean(),1)\n",
    "        score=0\n",
    "        combine_score=0\n",
    "        for other in other_groups:\n",
    "            other_idxs=samples[samples['Group']==other]['CODE_ID'].tolist()\n",
    "            other_values=Feature.loc[other_idxs,feature].values\n",
    "            df2.loc[i,other]=round(Feature.loc[other_idxs,feature].values.mean(),1)\n",
    "            t_stat, p_value = stats.ttest_ind(group_values,other_values, equal_var=False)\n",
    "            if (t_stat>0)&(p_value<0.01):\n",
    "                combine_score+=-np.log10(p_value)\n",
    "                score+=1\n",
    "                #df2.loc[i,other+\"_Score\"]=-np.log10(p_value)\n",
    "        df2.loc[i,group+\"_Score\"]=score\n",
    "        df2.loc[i,group+\"_SigScore\"]=combine_score\n",
    "    df2=df2[df2[group+'_Score']==len(other_groups)]\n",
    "    #df2=df2.sort_values(group+\"_SigScore\",ascending=False).reset_index(drop=True)\n",
    "    df2=df2.sort_values('Importance',ascending=False).reset_index(drop=True)\n",
    "    print(group, df2.shape)\n",
    "    df2.to_excel(\"D:\\\\18.TE_Evolution\\\\12.Analysis\\\\2024-08-16 TE Evolution\\\\Part3.Feature_Selection\\\\\"+group+\"_Features.xlsx\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e386b12f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
