{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "645f0b0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-09-25 13:14:34\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio import SeqUtils\n",
    "from Bio.Seq import Seq\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.set_option(\"display.float_format\",lambda x:'%.2f'%x) # 取消科学计数显示\n",
    "import time\n",
    "print (time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime()) )\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "from collections import Counter\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "26f04a3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "edta_dic={\"LTR/Copia\":\"Copia\",\"LTR/Gypsy\":\"Gypsy\",\"LTR/LINE\":\"LINE\",\"DNA/DTA\":\"hAT\", \"DNA/DTC\":\"CACTA\", 'LTR/unknown':'LTR Unknown',\n",
    "          \"Unknown\":\"Unknown\",'pararetrovirus':'Other','LINE/unknown':\"LINE\",'polinton':'Other', 'knob/knob180':'Other','LINE/L1':\"LINE\",\n",
    "          'subtelomere/4-12-1':'Other',\"MITE/DTA\":\"hAT\", \"MITE/DTC\":\"CACTA\", \"MITE/DTH\":\"PIF-Harbinger\", \"MITE/DTM\":\"Mutator\",\n",
    "          \"MITE/DTT\":\"Tc1_Mariner\",\"MITE/Helitron\":\"Helitron\",\n",
    "          \"DNA/DTH\":\"PIF-Harbinger\", \"DNA/DTM\":\"Mutator\", \"DNA/DTT\":\"Tc1_Mariner\",\"DNA/Helitron\":\"Helitron\",'TIR/unknown':\"TIR Unknown\"}\n",
    "TPSI_dic={'TY1_Copia':\"Copia\", 'gypsy':\"Gypsy\", 'LINE':\"LINE\",'ltr_Roo':\"LTR_Roo\", 'hAT':\"hAT\",'cacta':\"CACTA\", 'MuDR_A_B':\"Mutator\",\n",
    "          'mariner':\"Mariner\",'helitronORF':\"Helitron\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ccea9e48",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_edta_id(x):\n",
    "    values=x.split(\";\")\n",
    "    name=''\n",
    "    c=''\n",
    "    for value in values:\n",
    "        if value.startswith(\"Name=\"):\n",
    "            name=value.split('=')[1].strip()\n",
    "        if value.startswith(\"Classification=\"):\n",
    "            c=value.split('=')[1].strip()\n",
    "    return name+\"#\"+c\n",
    "def get_identity(x):\n",
    "    if 'homology' in x:\n",
    "        v=x.split(\"Identity=\")[1].split(\";\")[0]\n",
    "        if v=='NA':\n",
    "            return 0.5\n",
    "        else:\n",
    "            return float(v)\n",
    "    elif 'structural' in x:\n",
    "        #print(x)\n",
    "        if 'Identity=' in x:\n",
    "            v=x.split(\"Identity=\")[1].split(\";\")[0]\n",
    "            if v=='NA':\n",
    "                return 1\n",
    "            else:\n",
    "                return float(v)\n",
    "        elif \"identity=\" in x:\n",
    "            v=x.split(\"identity=\")[1].split(\";\")[0]\n",
    "            if v=='NA':\n",
    "                return 1\n",
    "            else:\n",
    "                return float(v)\n",
    "        else:\n",
    "            v=1\n",
    "        return v\n",
    "    else:\n",
    "        print(x)\n",
    "        return 0.5\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95116a1f",
   "metadata": {},
   "source": [
    "### Read the sample_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c4fddb86",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(558, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Specie_ID</th>\n",
       "      <th>Taxonomy_ID</th>\n",
       "      <th>Assembly_Accession</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>Genome_Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acer negundo</td>\n",
       "      <td>PBXX</td>\n",
       "      <td>4023</td>\n",
       "      <td>GCA_025594385.1</td>\n",
       "      <td>ASM2559438v1</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acer saccharum</td>\n",
       "      <td>QIBM</td>\n",
       "      <td>4024</td>\n",
       "      <td>GCA_030979865.1</td>\n",
       "      <td>UCONN_Acsa_2</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Acorus calamus</td>\n",
       "      <td>BUKK</td>\n",
       "      <td>4465</td>\n",
       "      <td>GCA_030737845.1</td>\n",
       "      <td>ASM3073784v1</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acorus gramineus</td>\n",
       "      <td>GJNO</td>\n",
       "      <td>55184</td>\n",
       "      <td>GCA_030737835.1</td>\n",
       "      <td>ASM3073783v1</td>\n",
       "      <td>NCBI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Actinidia chinensis</td>\n",
       "      <td>WWFJ</td>\n",
       "      <td>1590841</td>\n",
       "      <td>GCA_003024255.1</td>\n",
       "      <td>Red5_PS1_1.69.0</td>\n",
       "      <td>Ensemble</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Name Specie_ID  Taxonomy_ID Assembly_Accession  \\\n",
       "0         Acer negundo      PBXX         4023    GCA_025594385.1   \n",
       "1       Acer saccharum      QIBM         4024    GCA_030979865.1   \n",
       "2       Acorus calamus      BUKK         4465    GCA_030737845.1   \n",
       "3     Acorus gramineus      GJNO        55184    GCA_030737835.1   \n",
       "4  Actinidia chinensis      WWFJ      1590841    GCA_003024255.1   \n",
       "\n",
       "          Assembly Genome_Source  \n",
       "0     ASM2559438v1          NCBI  \n",
       "1     UCONN_Acsa_2          NCBI  \n",
       "2     ASM3073784v1          NCBI  \n",
       "3     ASM3073783v1          NCBI  \n",
       "4  Red5_PS1_1.69.0      Ensemble  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples=pd.read_excel(\"D:\\\\18.TE_Evolution\\\\00.Sample_Information\\\\Plant_558_Genome_Infor_this_project_2024-08-16.xlsx\").fillna(\"\")\n",
    "samples=samples.rename(columns={\"organism_name\":\"Name\"})\n",
    "print(samples.shape)\n",
    "samples.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a469befa",
   "metadata": {},
   "source": [
    "### Clean the results of DupGene_Finder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2c1c45a",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_path='D:\\\\18.TE_Evolution\\\\07.DupGene_Finder\\\\Clean\\\\'\n",
    "count=0\n",
    "for i in range(samples.shape[0]): # New Plant 13\n",
    "    specie_id=samples.loc[i,'Specie_ID']\n",
    "    specie_name=samples.loc[i,'Name'].replace(\" \",\"_\")\n",
    "    specie=samples.loc[i,'Name']#.replace(\" \",\"_\")\n",
    "    r=[]\n",
    "    file_path='D:\\\\18.TE_Evolution\\\\07.DupGene_Finder\\\\DupGene\\\\Plant_New50\\\\'\n",
    "    for query in os.listdir(file_path):\n",
    "        if query.startswith(specie_id):\n",
    "            files=os.listdir(file_path+query)\n",
    "            for f in files:\n",
    "                if f.endswith(\".pairs\"):\n",
    "                    df=pd.read_csv(file_path+query+\"//\"+f,sep='\\t')\n",
    "                    pair_type=f.split(\".\")[-2]\n",
    "                    cols=list(df.columns)\n",
    "                    df['Dup_Type']=pair_type\n",
    "                    df['Gene_1']=df[cols[0]]#.apply(lambda x:x.split(\"-RA\")[0])\n",
    "                    df['Gene_2']=df[cols[2]]#.apply(lambda x:x.split(\"-RA\")[0])\n",
    "                    df=df[['Gene_1','Gene_2','Dup_Type']]\n",
    "                    r.append(df)\n",
    "    r=pd.concat(r)\n",
    "    r['Pair']=r['Gene_1'].apply(lambda x:str(x))+\":\"+r['Gene_2'].apply(lambda x:str(x))\n",
    "    pair_dic={}\n",
    "    for p,p_df in r.groupby(\"Pair\"):\n",
    "        #print(p)\n",
    "        dup_type=Counter(p_df['Dup_Type'].tolist()).most_common(1)[0][0]\n",
    "        pair_dic[p]=dup_type\n",
    "    r=pd.DataFrame.from_dict(pair_dic,orient='index').reset_index()\n",
    "    r.columns=['Gene_Pair','Dup_Type']\n",
    "    count+=1\n",
    "    print(count,specie_id,r['Dup_Type'].value_counts())\n",
    "    r.to_csv(save_path+specie_id+\"_dup.csv\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50d83ab4",
   "metadata": {},
   "source": [
    "### Stat Results of DupGene_Finder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1b426665",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dup_Type</th>\n",
       "      <th>index</th>\n",
       "      <th>dispersed</th>\n",
       "      <th>transposed</th>\n",
       "      <th>proximal</th>\n",
       "      <th>tandem</th>\n",
       "      <th>wgd</th>\n",
       "      <th>Specie_ID</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>count</td>\n",
       "      <td>23454</td>\n",
       "      <td>6506.00</td>\n",
       "      <td>2666</td>\n",
       "      <td>2339</td>\n",
       "      <td>1998.00</td>\n",
       "      <td>PBXX</td>\n",
       "      <td>Acer negundo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>count</td>\n",
       "      <td>27443</td>\n",
       "      <td>8107.00</td>\n",
       "      <td>3729</td>\n",
       "      <td>3058</td>\n",
       "      <td>2087.00</td>\n",
       "      <td>QIBM</td>\n",
       "      <td>Acer saccharum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>count</td>\n",
       "      <td>30068</td>\n",
       "      <td>5592.00</td>\n",
       "      <td>3048</td>\n",
       "      <td>2229</td>\n",
       "      <td>25317.00</td>\n",
       "      <td>BUKK</td>\n",
       "      <td>Acorus calamus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>count</td>\n",
       "      <td>17079</td>\n",
       "      <td>5823.00</td>\n",
       "      <td>1487</td>\n",
       "      <td>1598</td>\n",
       "      <td>3345.00</td>\n",
       "      <td>GJNO</td>\n",
       "      <td>Acorus gramineus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>count</td>\n",
       "      <td>22660</td>\n",
       "      <td>4925.00</td>\n",
       "      <td>368</td>\n",
       "      <td>1528</td>\n",
       "      <td>19764.00</td>\n",
       "      <td>WWFJ</td>\n",
       "      <td>Actinidia chinensis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dup_Type  index  dispersed  transposed  proximal  tandem      wgd Specie_ID  \\\n",
       "0         count      23454     6506.00      2666    2339  1998.00      PBXX   \n",
       "0         count      27443     8107.00      3729    3058  2087.00      QIBM   \n",
       "0         count      30068     5592.00      3048    2229 25317.00      BUKK   \n",
       "0         count      17079     5823.00      1487    1598  3345.00      GJNO   \n",
       "0         count      22660     4925.00       368    1528 19764.00      WWFJ   \n",
       "\n",
       "Dup_Type                 Name  \n",
       "0                Acer negundo  \n",
       "0              Acer saccharum  \n",
       "0              Acorus calamus  \n",
       "0            Acorus gramineus  \n",
       "0         Actinidia chinensis  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "save_path='D://18.TE_Evolution//07.DupGene_Finder//Clean//'\n",
    "count=0\n",
    "R=[]\n",
    "for i in range(samples.shape[0]):\n",
    "    if i%100==0:\n",
    "        print(i)\n",
    "    specie_id=samples.loc[i,'Specie_ID']\n",
    "    specie_name=samples.loc[i,'Name']#.replace(\" \",\"_\")\n",
    "    specie=samples.loc[i,'Name']#.replace(\" \",\"_\")\n",
    "    #sample=samples.loc[i,'Assembly'] #ZZAX_protein.fa\n",
    "    dup=pd.read_csv(save_path+specie_id+\"_dup.csv\")\n",
    "    df0=pd.DataFrame(dup['Dup_Type'].value_counts()).T.reset_index()\n",
    "    df0['Specie_ID']=specie_id\n",
    "    df0['Name']=specie_name\n",
    "    R.append(df0)\n",
    "R=pd.concat(R)\n",
    "R.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c93a081a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "transposed 3688493 10.94\n",
      "proximal 1015256 3.01\n",
      "tandem 4883972 14.48\n",
      "wgd 5457283 16.18\n",
      "dispersed 18677688 55.39\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dup_Type</th>\n",
       "      <th>Specie_ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>transposed</th>\n",
       "      <th>proximal</th>\n",
       "      <th>tandem</th>\n",
       "      <th>wgd</th>\n",
       "      <th>dispersed</th>\n",
       "      <th>Total</th>\n",
       "      <th>transposed(%)</th>\n",
       "      <th>proximal(%)</th>\n",
       "      <th>tandem(%)</th>\n",
       "      <th>wgd(%)</th>\n",
       "      <th>dispersed(%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PBXX</td>\n",
       "      <td>Acer negundo</td>\n",
       "      <td>6506.00</td>\n",
       "      <td>2666</td>\n",
       "      <td>2339</td>\n",
       "      <td>1998.00</td>\n",
       "      <td>23454</td>\n",
       "      <td>36963.00</td>\n",
       "      <td>17.60</td>\n",
       "      <td>7.21</td>\n",
       "      <td>6.33</td>\n",
       "      <td>5.41</td>\n",
       "      <td>63.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>QIBM</td>\n",
       "      <td>Acer saccharum</td>\n",
       "      <td>8107.00</td>\n",
       "      <td>3729</td>\n",
       "      <td>3058</td>\n",
       "      <td>2087.00</td>\n",
       "      <td>27443</td>\n",
       "      <td>44424.00</td>\n",
       "      <td>18.25</td>\n",
       "      <td>8.39</td>\n",
       "      <td>6.88</td>\n",
       "      <td>4.70</td>\n",
       "      <td>61.78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BUKK</td>\n",
       "      <td>Acorus calamus</td>\n",
       "      <td>5592.00</td>\n",
       "      <td>3048</td>\n",
       "      <td>2229</td>\n",
       "      <td>25317.00</td>\n",
       "      <td>30068</td>\n",
       "      <td>66254.00</td>\n",
       "      <td>8.44</td>\n",
       "      <td>4.60</td>\n",
       "      <td>3.36</td>\n",
       "      <td>38.21</td>\n",
       "      <td>45.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GJNO</td>\n",
       "      <td>Acorus gramineus</td>\n",
       "      <td>5823.00</td>\n",
       "      <td>1487</td>\n",
       "      <td>1598</td>\n",
       "      <td>3345.00</td>\n",
       "      <td>17079</td>\n",
       "      <td>29332.00</td>\n",
       "      <td>19.85</td>\n",
       "      <td>5.07</td>\n",
       "      <td>5.45</td>\n",
       "      <td>11.40</td>\n",
       "      <td>58.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WWFJ</td>\n",
       "      <td>Actinidia chinensis</td>\n",
       "      <td>4925.00</td>\n",
       "      <td>368</td>\n",
       "      <td>1528</td>\n",
       "      <td>19764.00</td>\n",
       "      <td>22660</td>\n",
       "      <td>49245.00</td>\n",
       "      <td>10.00</td>\n",
       "      <td>0.75</td>\n",
       "      <td>3.10</td>\n",
       "      <td>40.13</td>\n",
       "      <td>46.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dup_Type Specie_ID                 Name  transposed  proximal  tandem  \\\n",
       "0             PBXX         Acer negundo     6506.00      2666    2339   \n",
       "0             QIBM       Acer saccharum     8107.00      3729    3058   \n",
       "0             BUKK       Acorus calamus     5592.00      3048    2229   \n",
       "0             GJNO     Acorus gramineus     5823.00      1487    1598   \n",
       "0             WWFJ  Actinidia chinensis     4925.00       368    1528   \n",
       "\n",
       "Dup_Type      wgd  dispersed    Total  transposed(%)  proximal(%)  tandem(%)  \\\n",
       "0         1998.00      23454 36963.00          17.60         7.21       6.33   \n",
       "0         2087.00      27443 44424.00          18.25         8.39       6.88   \n",
       "0        25317.00      30068 66254.00           8.44         4.60       3.36   \n",
       "0         3345.00      17079 29332.00          19.85         5.07       5.45   \n",
       "0        19764.00      22660 49245.00          10.00         0.75       3.10   \n",
       "\n",
       "Dup_Type  wgd(%)  dispersed(%)  \n",
       "0           5.41         63.45  \n",
       "0           4.70         61.78  \n",
       "0          38.21         45.38  \n",
       "0          11.40         58.23  \n",
       "0          40.13         46.01  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "R=R[['Specie_ID','Name','transposed', 'proximal', 'tandem', 'wgd','dispersed']]\n",
    "R['Total']=R[['transposed', 'proximal', 'tandem', 'wgd','dispersed']].sum(axis=1)\n",
    "for col in ['transposed', 'proximal', 'tandem', 'wgd','dispersed']:\n",
    "    R[col+\"(%)\"]=R[col]/R[\"Total\"]*100\n",
    "    print(col,int(R[col].sum()), round(int(R[col].sum())/int(R['Total'].sum())*100,2))\n",
    "R.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "771bdad2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33722692\n"
     ]
    }
   ],
   "source": [
    "print(int(R[\"Total\"].sum()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2fd7c493",
   "metadata": {},
   "outputs": [],
   "source": [
    "R.to_excel(\"D://18.TE_Evolution//07.DupGene_Finder//Plant_558_DupgeneFinder_GeneDup_Stats.xlsx\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba76c90d",
   "metadata": {},
   "source": [
    "### Filter Results of DupGene Finder Using TE information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a165e6f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install python-louvain\n",
    "file_path='D:\\\\18.TE_Evolution\\\\07.DupGene_Finder\\\\Clean\\\\'\n",
    "import pandas as pd\n",
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt\n",
    "import community  # 需要安装 python-louvain 库 !pip install python-louvain\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.metrics.pairwise import cosine_similarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f84f3c5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_region_te_seqs(region):\n",
    "    c1=region.split(\":\")[0]\n",
    "    s1=int(region.split(\":\")[1].split(\"-\")[0])\n",
    "    e1=int(region.split(\":\")[1].split(\"-\")[1])\n",
    "    if c1 in TE_dic:\n",
    "        te=TE_dic[c1]\n",
    "        df=te[(te[\"Start\"]>=s1)&(te['End']<=e1)]\n",
    "        if df.shape[0]>0:\n",
    "            seq1=df['TE_Type'].tolist()\n",
    "            return \" \".join(seq1)\n",
    "    return ''\n",
    "def get_region_tepep_seqs(region):\n",
    "    c1=region.split(\":\")[0]\n",
    "    s1=int(region.split(\":\")[1].split(\"-\")[0])\n",
    "    e1=int(region.split(\":\")[1].split(\"-\")[1])\n",
    "    if c1 in TEpep_dic:\n",
    "        tepep=TEpep_dic[c1]\n",
    "        df=tepep[(tepep[\"Start\"]>=s1)&(tepep['End']<=e1)]\n",
    "        if df.shape[0]>0:\n",
    "            seq1=df['TE_Type'].tolist()\n",
    "            return \" \".join(seq1)\n",
    "    return ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1e71d5d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_Gene_Pair_COS_Sim(pair):\n",
    "    seq1=pair.split(\":\")[0]\n",
    "    seq2=pair.split(\":\")[1]\n",
    "    if (seq1 !=' ') &(seq2!=' '):\n",
    "        vectorizer = TfidfVectorizer()\n",
    "        tfidf_matrix = vectorizer.fit_transform([seq1, seq2])\n",
    "        # 计算余弦相似度\n",
    "        cosine_sim = cosine_similarity(tfidf_matrix[0:1], tfidf_matrix[1:2])\n",
    "        return str(cosine_sim[0][0])+\":\"+seq1+\"; \"+seq2\n",
    "    else:\n",
    "        return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "60bad731",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Specie_ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>transposed</th>\n",
       "      <th>proximal</th>\n",
       "      <th>tandem</th>\n",
       "      <th>wgd</th>\n",
       "      <th>dispersed</th>\n",
       "      <th>Total</th>\n",
       "      <th>transposed(%)</th>\n",
       "      <th>proximal(%)</th>\n",
       "      <th>tandem(%)</th>\n",
       "      <th>wgd(%)</th>\n",
       "      <th>dispersed(%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PBXX</td>\n",
       "      <td>Acer negundo</td>\n",
       "      <td>6506.00</td>\n",
       "      <td>2666</td>\n",
       "      <td>2339</td>\n",
       "      <td>1998.00</td>\n",
       "      <td>23454</td>\n",
       "      <td>36963</td>\n",
       "      <td>17.60</td>\n",
       "      <td>7.21</td>\n",
       "      <td>6.33</td>\n",
       "      <td>5.41</td>\n",
       "      <td>63.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>QIBM</td>\n",
       "      <td>Acer saccharum</td>\n",
       "      <td>8107.00</td>\n",
       "      <td>3729</td>\n",
       "      <td>3058</td>\n",
       "      <td>2087.00</td>\n",
       "      <td>27443</td>\n",
       "      <td>44424</td>\n",
       "      <td>18.25</td>\n",
       "      <td>8.39</td>\n",
       "      <td>6.88</td>\n",
       "      <td>4.70</td>\n",
       "      <td>61.78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BUKK</td>\n",
       "      <td>Acorus calamus</td>\n",
       "      <td>5592.00</td>\n",
       "      <td>3048</td>\n",
       "      <td>2229</td>\n",
       "      <td>25317.00</td>\n",
       "      <td>30068</td>\n",
       "      <td>66254</td>\n",
       "      <td>8.44</td>\n",
       "      <td>4.60</td>\n",
       "      <td>3.36</td>\n",
       "      <td>38.21</td>\n",
       "      <td>45.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GJNO</td>\n",
       "      <td>Acorus gramineus</td>\n",
       "      <td>5823.00</td>\n",
       "      <td>1487</td>\n",
       "      <td>1598</td>\n",
       "      <td>3345.00</td>\n",
       "      <td>17079</td>\n",
       "      <td>29332</td>\n",
       "      <td>19.85</td>\n",
       "      <td>5.07</td>\n",
       "      <td>5.45</td>\n",
       "      <td>11.40</td>\n",
       "      <td>58.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WWFJ</td>\n",
       "      <td>Actinidia chinensis</td>\n",
       "      <td>4925.00</td>\n",
       "      <td>368</td>\n",
       "      <td>1528</td>\n",
       "      <td>19764.00</td>\n",
       "      <td>22660</td>\n",
       "      <td>49245</td>\n",
       "      <td>10.00</td>\n",
       "      <td>0.75</td>\n",
       "      <td>3.10</td>\n",
       "      <td>40.13</td>\n",
       "      <td>46.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Specie_ID                 Name  transposed  proximal  tandem      wgd  \\\n",
       "0      PBXX         Acer negundo     6506.00      2666    2339  1998.00   \n",
       "1      QIBM       Acer saccharum     8107.00      3729    3058  2087.00   \n",
       "2      BUKK       Acorus calamus     5592.00      3048    2229 25317.00   \n",
       "3      GJNO     Acorus gramineus     5823.00      1487    1598  3345.00   \n",
       "4      WWFJ  Actinidia chinensis     4925.00       368    1528 19764.00   \n",
       "\n",
       "   dispersed  Total  transposed(%)  proximal(%)  tandem(%)  wgd(%)  \\\n",
       "0      23454  36963          17.60         7.21       6.33    5.41   \n",
       "1      27443  44424          18.25         8.39       6.88    4.70   \n",
       "2      30068  66254           8.44         4.60       3.36   38.21   \n",
       "3      17079  29332          19.85         5.07       5.45   11.40   \n",
       "4      22660  49245          10.00         0.75       3.10   40.13   \n",
       "\n",
       "   dispersed(%)  \n",
       "0         63.45  \n",
       "1         61.78  \n",
       "2         45.38  \n",
       "3         58.23  \n",
       "4         46.01  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stat=pd.read_excel(\"D://18.TE_Evolution//07.DupGene_Finder//Plant_558_DupgeneFinder_GeneDup_Stats.xlsx\")\n",
    "stat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "866f1428",
   "metadata": {},
   "outputs": [],
   "source": [
    "window=2000\n",
    "for i in range(samples.shape[0]):\n",
    "    specie_id=samples.loc[i,'Specie_ID']\n",
    "    if specie_id+\"_dup.csv\" in os.listdir(file_path):\n",
    "        genedup=pd.read_csv(file_path+specie_id+\"_dup.csv\")\n",
    "        genedup=genedup[genedup['Dup_Type'].apply(lambda x:1 if x in ['dispersed','transposed'] else 0)==1].reset_index(drop=True)\n",
    "        if i==68:\n",
    "            genedup['Source']=genedup['Gene_Pair'].apply(lambda x:\":\".join(x.split(\":\")[:2]))\n",
    "            genedup['Target']=genedup['Gene_Pair'].apply(lambda x:\":\".join(x.split(\":\")[-2:]))\n",
    "        else:\n",
    "            genedup['Source']=genedup['Gene_Pair'].apply(lambda x:x.split(\":\")[0])\n",
    "            genedup['Target']=genedup['Gene_Pair'].apply(lambda x:x.split(\":\")[1])\n",
    "        dupgenes={}\n",
    "        for gene in genedup['Source'].unique():\n",
    "            dupgenes[gene]=1\n",
    "        for gene in genedup['Target'].unique():\n",
    "            dupgenes[gene]=1\n",
    "        bed_path='D:\\\\18.TE_Evolution\\\\06.Bed\\\\'\n",
    "        TE=pd.read_csv(bed_path+\"TE_Bed\\\\\"+specie_id+\"_EDTA.bed\",sep='\\t')\n",
    "        TE_dic={}\n",
    "        for chrom,c_df in TE.groupby(\"#CHROM\"):\n",
    "            TE_dic[str(chrom)]=c_df\n",
    "        TEpep=pd.read_csv(bed_path+\"TEpep_Bed\\\\\"+specie_id+\"_TPSI.bed\",sep='\\t')\n",
    "        TEpep_dic={}\n",
    "        for chrom,c_df in TEpep.groupby(\"#CHROM\"):\n",
    "            TEpep_dic[str(chrom)]=c_df\n",
    "        del TE\n",
    "        del TEpep\n",
    "        gene_bed=pd.read_csv(bed_path+\"Gene_Bed_with_Strand\\\\\"+specie_id+\"_Gene.bed\",sep='\\t')\n",
    "        gene_bed['Gene']=gene_bed['Gene'].apply(lambda x:str(x))\n",
    "        gene_bed=gene_bed[gene_bed['Gene'].apply(lambda x:1 if x in dupgenes else 0)==1].reset_index(drop=True)\n",
    "        #gene_bed['Start']=gene_bed['Start'].apply(lambda x: x-10000 if x>10000 else x)\n",
    "        gene_bed['Start']=gene_bed['Start'].apply(lambda x: x-window if x>window else x)\n",
    "        if i==195:\n",
    "            #gene_bed['End']=gene_bed['End'].apply(lambda x:int(str(x).split(\" \")[0])+10000)\n",
    "            gene_bed['End']=gene_bed['End'].apply(lambda x:int(str(x).split(\" \")[0])+window)\n",
    "        else:\n",
    "            gene_bed['End']=gene_bed['End']+window\n",
    "        gene_bed['Region']=gene_bed['#CHROM'].apply(lambda x:str(x))+\":\"+gene_bed['Start'].apply(lambda x:str(x))+'-'+gene_bed['End'].apply(lambda x:str(x))\n",
    "        gene_bed['TE_Seq']=gene_bed['Region'].apply(get_region_te_seqs)\n",
    "        gene_bed['TEpep_Seq']=gene_bed['Region'].apply(get_region_tepep_seqs)\n",
    "        print(\"Raw:\\t\",specie_id,genedup.shape)\n",
    "        dupgenes={}\n",
    "        for j in range(gene_bed.shape[0]):\n",
    "            dupgenes[gene_bed.loc[j,'Gene']]={}\n",
    "            dupgenes[gene_bed.loc[j,'Gene']]['TE_Seq']=gene_bed.loc[j,'TE_Seq']\n",
    "            dupgenes[gene_bed.loc[j,'Gene']]['TEpep_Seq']=gene_bed.loc[j,'TEpep_Seq']\n",
    "        genedup['Seq1']=genedup['Source'].apply(lambda x:str(dupgenes[x]['TE_Seq'])+\" \"+str(dupgenes[x]['TEpep_Seq']) if x in dupgenes else '')\n",
    "        genedup['Seq2']=genedup['Target'].apply(lambda x:str(dupgenes[x]['TE_Seq'])+\" \"+str(dupgenes[x]['TEpep_Seq']) if x in dupgenes else '')\n",
    "        genedup=genedup[(genedup[\"Seq1\"]!=' ')&(genedup[\"Seq2\"]!=\" \")]\n",
    "        genedup['Seq']=genedup['Seq1']+\":\"+genedup['Seq2']\n",
    "        genedup['Cos_Sim']=genedup['Seq'].apply(get_Gene_Pair_COS_Sim)\n",
    "        genedup['Cos_Sim2']=genedup['Cos_Sim'].fillna(\"0:\").apply(lambda x:float(x.split(\":\")[0]))\n",
    "        genedup=genedup[genedup['Cos_Sim2']>0.5]\n",
    "        print(specie_id,'Round 2 \\t',genedup.shape[0])#r0.shape[0],\n",
    "        genedup.to_csv('D:\\\\18.TE_Evolution\\\\07.DupGene_Finder\\\\Filtered(2K)\\\\'+specie_id+\"_Dups.csv\",index=False)\n",
    "        df = genedup[['Source','Target']]\n",
    "        # 创建有向图\n",
    "        G = nx.from_pandas_edgelist(df, 'Source', 'Target', create_using=nx.DiGraph())\n",
    "            # 找到所有的连接模块\n",
    "            # 使用社区检测算法（例如，Louvain 算法）\n",
    "        partition = community.best_partition(G.to_undirected())\n",
    "            # 将模块信息存储在 DataFrame 中\n",
    "        modules = pd.DataFrame(list(partition.items()), columns=['Node', 'Module'])\n",
    "        modules.to_csv('D:\\\\18.TE_Evolution\\\\07.DupGene_Finder\\\\Filtered(2K)\\\\'+specie_id+\"_Dups_Modules.csv\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54ab36f4",
   "metadata": {},
   "source": [
    "### Stat Dup Gene List"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37b58a93",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path='D:\\\\18.TE_Evolution\\\\07.DupGene_Finder\\\\Filtered(2K)\\\\'\n",
    "save_path='D:\\\\18.TE_Evolution\\\\07.DupGene_Finder\\\\TE_Dup_Genes(2K)\\\\Overall\\\\'\n",
    "total_count=0\n",
    "for i in range(samples.shape[0]):\n",
    "    specie_id=samples.loc[i,'Specie_ID']\n",
    "    dup_genes=[]\n",
    "    if specie_id+\"_Dups_Modules.csv\" in os.listdir(file_path):\n",
    "            df=pd.read_csv(file_path+specie_id+\"_Dups_Modules.csv\")\n",
    "            dup_genes+=df['Node'].tolist()\n",
    "            #df=df[df[3]==df[10]] # 方向相同\n",
    "    if dup_genes!=[]:\n",
    "        with open(save_path+specie_id+\".TE_DupGenes.txt\",'w') as f:\n",
    "            for gene in list(set(dup_genes)):\n",
    "                f.write(str(gene)+'\\n')\n",
    "                total_count+=1\n",
    "            f.close()\n",
    "total_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6adfed3b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58f35b68",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_overlap(x):\n",
    "    seq1=x.split(\":\")[0].split(\" \")\n",
    "    seq2=x.split(\":\")[0].split(\" \")\n",
    "    overlap=list(set(seq1).intersection(set(seq2)))\n",
    "    if '' in overlap:\n",
    "        overlap.remove(\"\")\n",
    "    return \" | \".join(overlap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40b6b4ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "window='2K'\n",
    "file_path='D:\\\\18.TE_Evolution\\\\07.DupGene_Finder\\\\Filtered('+window+')\\\\'\n",
    "save_path='D:\\\\18.TE_Evolution\\\\07.DupGene_Finder\\\\TE_Dup_Genes('+window+')\\\\Super_Family\\\\'\n",
    "tepep_path='D:\\\\18.TE_Evolution\\\\05.TEpep\\\\Gene_TEpep_Genelist\\\\Super_Family\\\\'\n",
    "families=['Copia','Gypsy', 'CACTA', 'hAT','Tc1_Mariner','Mutator', 'Helitron', 'PIF-Harbinger']\n",
    "Dup_genes_stat={}\n",
    "te_gene_dic={}\n",
    "tepep_gene_dic={}\n",
    "for i in range(samples.shape[0]):\n",
    "    if i%10==0:\n",
    "        print(i)\n",
    "    specie_id=samples.loc[i,'Specie_ID']\n",
    "    for f in os.listdir(tepep_path):\n",
    "        if specie_id in f:\n",
    "            family=f.split(specie_id)[0][:-1]\n",
    "            for gene in [gene.strip() for gene in open(tepep_path+f,'r').readlines()]:\n",
    "                tepep_gene_dic[gene]=family\n",
    "    \n",
    "    if specie_id+\".TE_vs_Gene.bed\" in os.listdir(\"D:\\\\18.TE_Evolution\\\\04.TE\\Gene_TE\\\\\"):\n",
    "        te=pd.read_csv(\"D:\\\\18.TE_Evolution\\\\04.TE\\Gene_TE\\\\\"+specie_id+\".TE_vs_Gene.bed\",sep='\\t',header=None)\n",
    "        for g,g_df in te.groupby(13):\n",
    "            cs=Counter(g_df[7].tolist())\n",
    "            te_gene_dic[g]=cs.most_common(1)[0][0]\n",
    "            break\n",
    "    df=pd.read_csv(file_path+specie_id+\"_Dups.csv\")\n",
    "    df['Seq']=df[\"Seq1\"]+\":\"+df[\"Seq2\"]\n",
    "    df['Overlap']=df[\"Seq\"].apply(get_overlap)\n",
    "    df0=df[df['Overlap']!=''].reset_index(drop=True)\n",
    "    df0=df0[['Source','Target','Overlap']]\n",
    "    df0['Source_TE_Class']=df0['Source'].apply(lambda x:te_gene_dic[x] if x in te_gene_dic else '')\n",
    "    df0['Target_TE_Class']=df0['Target'].apply(lambda x:te_gene_dic[x] if x in te_gene_dic else '')\n",
    "    df0['Source_TEpep_Class']=df0['Source'].apply(lambda x:tepep_gene_dic[x] if x in tepep_gene_dic else '')\n",
    "    df0['Target_TEpep_Class']=df0['Target'].apply(lambda x:tepep_gene_dic[x] if x in tepep_gene_dic else '')\n",
    "    for i in range(df0.shape[0]):\n",
    "        if df0.loc[i,'Source_TE_Class']!='':\n",
    "            if df0.loc[i,'Source_TE_Class']==df0.loc[i,'Target_TE_Class']:\n",
    "                df0.loc[i,'Overlap']=df0.loc[i,'Source_TE_Class']\n",
    "        elif df0.loc[i,'Source_TEpep_Class']!='':\n",
    "            if df0.loc[i,'Source_TEpep_Class']==df0.loc[i,'Target_TEpep_Class']:\n",
    "                df0.loc[i,'Overlap']=df0.loc[i,'Source_TEpep_Class']\n",
    "    df1=df0[(df0['Overlap']==df0['Source_TE_Class'])|(df0['Overlap']==df0['Source_TEpep_Class'])]\n",
    "    df2=df0[(df0['Overlap']!=df0['Source_TE_Class'])&(df0['Overlap']!=df0['Source_TEpep_Class'])]\n",
    "    df0=pd.concat([df1,df2])\n",
    "            \n",
    "    for family in families:\n",
    "        if family not in Dup_genes_stat:\n",
    "            Dup_genes_stat[family]=0\n",
    "        df1=df0[df0['Overlap'].apply(lambda x:1 if family in x else 0)==1]\n",
    "        family_dup_genes=list(set(df1['Source'].tolist()+df1['Target'].tolist()))\n",
    "        Dup_genes_stat[family]+=len(family_dup_genes)\n",
    "        with open(save_path+family+\"_\"+specie_id+\".TE_DupGenes.txt\",'w') as f:\n",
    "            for gene in family_dup_genes:\n",
    "                f.write(str(gene)+'\\n')\n",
    "            f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "108d7c99",
   "metadata": {},
   "outputs": [],
   "source": [
    "df0=pd.DataFrame.from_dict(Dup_genes_stat,orient='index').reset_index()\n",
    "df0.columns=['TE_Super_Family','DupGene_Number']\n",
    "df0=df0.sort_values(\"DupGene_Number\",ascending=False)\n",
    "df0.to_excel(\"D:\\\\18.TE_Evolution\\\\07.DupGene_Finder\\\\TE_Dup_Genes(\"+window+\")\\\\Super_Family_Stats(\"+window+\").xlsx\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "546f5e68",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
